clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordBatch.java

```diff
@@ -86,33 +86,32 @@ public FutureRecordMetadata tryAppend(long timestamp, byte[] key, byte[] value,
      * Complete the request
      * 
      * @param baseOffset The base offset of the messages assigned by the server
-     * @param timestamp The timestamp returned by the broker.
+     * @param responseTimestamp The timestamp returned by the broker.
      * @param exception The exception that occurred (or null if the request was successful)
      */
-    public void done(long baseOffset, long timestamp, RuntimeException exception) {
+    public void done(long baseOffset, long responseTimestamp, RuntimeException exception) {
         log.trace("Produced messages to topic-partition {} with base offset offset {} and error: {}.",
                   topicPartition,
                   baseOffset,
                   exception);
+
+        // set the future before invoking the callbacks as we rely on its state for the `onCompletion` call
+        produceFuture.set(topicPartition, baseOffset, responseTimestamp, exception);
+
         // execute callbacks
         for (Thunk thunk : thunks) {
             try {
                 if (exception == null) {
-                    // If the timestamp returned by server is NoTimestamp, that means CreateTime is used. Otherwise LogAppendTime is used.
-                    RecordMetadata metadata = new RecordMetadata(this.topicPartition,  baseOffset, thunk.future.relativeOffset(),
-                                                                 timestamp == Record.NO_TIMESTAMP ? thunk.future.timestamp() : timestamp,
-                                                                 thunk.future.checksum(),
-                                                                 thunk.future.serializedKeySize(),
-                                                                 thunk.future.serializedValueSize());
+                    RecordMetadata metadata = thunk.future.value();
```

Comment:
What makes this difficult to follow is that `value()` depends indirectly on the fields set in `produceFuture.set()` above. I think this is ok here, but I'm wondering if a separate refactor could do something like this:

1. Pull `ProduceRequestResult` out of `FutureRecordMetadata`.
2. Pull the latch out of `ProduceRequestResult` and into `RecordBatch`. 
3. Each instance of `FutureRecordMetadata` can have a reference to the latch instead of `ProduceRequestResult`
4. Make `ProduceRequestResult` immutable and only construct it when the result is ready.
5. Add a `FutureRecordMetadata.complete(ProduceRequestResult)`.